<a href="https://colab.research.google.com/github/adamazanos/challenge_1_binar/blob/main/challenge_1_v2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Tugas Binar GOLD 1**

---

Jadi targetnya, diminta buat API yang bisa prediksi kalimat abuse di twitter berbahasa indonesia

PAKE RUMUSAN 
**3 TAHAP**

A. IMPORT DATA

B. BERESIN DAN CLEANSING DATA

1.   DI BERESI FORMATNYA JADI KECIL SEMUA
2.   PAKAI DICT
3.   PAKAI REGEX HAPUS HAPUSIN YANG KOTOR

C. URUSAN API

# A. URUSAN DATA

1. Kite download dan import data dari kaggle di link
(https://www.kaggle.com/datasets/ilhamfp31/indonesian-abusive-and-hate-speech-twitter-text)
2. install library (atau plugin) yang dibutuhin

*   numpy
*   pandas
*   dataframe

3. datanya teksnya setelah kita opname, kita cek ricek, bagian pala ama buntut, banyak nya kalimat atau kata macam apa.

# i INSTALL LIBRARY


**1. INSTALL SEMUA LIBRARY AMUNISI BUAT BACA DAN RECEK TEKS**



In [ ]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns

# ii BACA CSV TUGAS

2. BACA FILE CSV DARI TUGAS

Filenya di masukin ke google drive pribadi dan di read.

*pakai logika alur di https://www.kaggle.com/code/corneliustantius/indonesia-hatespeech-recognition-main-notebook*

In [ ]:
data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/01. BINAR CHALLANGE 01/abuse word twitter data/data.csv', encoding='latin-1')
alay_dict = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/01. BINAR CHALLANGE 01/abuse word twitter data/new_kamusalay.csv', names = ['original', 'replacement'], encoding='latin-1')
abusive_dict = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/01. BINAR CHALLANGE 01/abuse word twitter data/abusive.csv', encoding='latin-1')

# iii CEK DATA MANUAL

**3. KITA CEK RICEK DATANYA**
 
 a. cek seluruh tabel berapa banyak kolom dan baris
 
 b. cek data pala dan ekor

11 data awal dan 11 data akhir 

In [ ]:
data

Tweet  HS  Abusive  \
0      - disaat semua cowok berusaha melacak perhatia...   1        1   
1      RT USER: USER siapa yang telat ngasih tau elu?...   0        1   
2      41. Kadang aku berfikir, kenapa aku tetap perc...   0        0   
3      USER USER AKU ITU AKU\n\nKU TAU MATAMU SIPIT T...   0        0   
4      USER USER Kaum cebong kapir udah keliatan dong...   1        1   
...                                                  ...  ..      ...   
13164  USER jangan asal ngomong ndasmu. congor lu yg ...   1        1   
13165                       USER Kasur mana enak kunyuk'   0        1   
13166  USER Hati hati bisu :( .g\n\nlagi bosan huft \...   0        0   
13167  USER USER USER USER Bom yang real mudah terdet...   0        0   
13168  USER Mana situ ngasih(": itu cuma foto ya kuti...   1        1   

       HS_Individual  HS_Group  HS_Religion  HS_Race  HS_Physical  HS_Gender  \
0                  1         0            0        0            0          0   
1                  0         0            0        0            0          0   
2                  0         0            0        0            0          0   
3                  0         0            0        0            0          0   
4                  0         1            1        0            0          0   
...              ...       ...          ...      ...          ...        ...   
13164              1         0            0        0            1          0   
13165              0         0            0        0            0          0   
13166              0         0            0        0            0          0   
13167              0         0            0        0            0          0   
13168              1         0            0        0            0          0   

       HS_Other  HS_Weak  HS_Moderate  HS_Strong  
0             1        1            0          0  
1             0        0            0          0  
2             0        0            0          0  
3             0        0            0          0  
4             0        0            1          0  
...         ...      ...          ...        ...  
13164         0        1            0          0  
13165         0        0            0          0  
13166         0        0            0          0  
13167         0        0            0          0  
13168         1        1            0          0  

[13169 rows x 13 columns]

In [ ]:
data.head(25)

Tweet  HS  Abusive  \
0   - disaat semua cowok berusaha melacak perhatia...   1        1   
1   RT USER: USER siapa yang telat ngasih tau elu?...   0        1   
2   41. Kadang aku berfikir, kenapa aku tetap perc...   0        0   
3   USER USER AKU ITU AKU\n\nKU TAU MATAMU SIPIT T...   0        0   
4   USER USER Kaum cebong kapir udah keliatan dong...   1        1   
5   USER Ya bani taplak dkk \xf0\x9f\x98\x84\xf0\x...   1        1   
6   deklarasi pilkada 2018 aman dan anti hoax warg...   0        0   
7   Gue baru aja kelar re-watch Aldnoah Zero!!! pa...   0        1   
8   Nah admin belanja satu lagi port terbaik nak m...   0        0   
9                       USER Enak lg klo smbil ngewe'   0        1   
10  Setidaknya gw punya jari tengah buat lu, sebel...   1        1   
11  USER USER USER USER BANCI KALENG MALU GA BISA ...   1        1   
12  Kalo belajar ekonomi mestinya jago memprivatis...   0        0   
13  Aktor huruhara 98 Prabowo S ingin lengserkan p...   1        0   
14  USER Bu guru enakan jadi jablay atau guru esde...   1        1   
15  USER USER USER USER USER USER Lawan bicara gw ...   1        1   
16        Belakangan ini kok fikiran ampas banget ya'   0        0   
17  Ari sarua beki mah repeh monyet\xf0\x9f\x98\x8...   0        1   
18  Jadi cowo itu harus Gantle kalo ga Gantle itu ...   1        1   
19                USER Slga mnr bom \xf0\x9f\x98\x82'   0        0   
20  Asw ya tapi gua jarang ngambek, tacut wkkwkwkw...   0        1   
21  USER kalo kamu noob pasti peluang disakitin ny...   0        0   
22  USER Joko Widodo dinilai sebagai presiden terl...   0        0   
23  PELAJAR SMA KEC BILAH HILIR DEKLARASI ANTI HOA...   0        0   
24  Bandara Udara Internasional Kertajati dibangun...   0        0   

    HS_Individual  HS_Group  HS_Religion  HS_Race  HS_Physical  HS_Gender  \
0               1         0            0        0            0          0   
1               0         0            0        0            0          0   
2               0         0            0        0            0          0   
3               0         0            0        0            0          0   
4               0         1            1        0            0          0   
5               0         1            0        0            0          0   
6               0         0            0        0            0          0   
7               0         0            0        0            0          0   
8               0         0            0        0            0          0   
9               0         0            0        0            0          0   
10              1         0            0        0            0          0   
11              1         0            0        0            0          1   
12              0         0            0        0            0          0   
13              1         0            0        0            0          0   
14              1         0            0        0            0          1   
15              1         0            1        0            0          0   
16              0         0            0        0            0          0   
17              0         0            0        0            0          0   
18              0         1            0        0            0          1   
19              0         0            0        0            0          0   
20              0         0            0        0            0          0   
21              0         0            0        0            0          0   
22              0         0            0        0            0          0   
23              0         0            0        0            0          0   
24              0         0            0        0            0          0   

    HS_Other  HS_Weak  HS_Moderate  HS_Strong  
0          1        1            0          0  
1          0        0            0          0  
2          0        0            0          0  
3          0        0           

In [ ]:
data.tail(11)

Tweet  HS  Abusive  \
13158  Maaf sianu mah cocoknya sama mustafa kemal alm...   0        0   
13159  Ada 2 buku aktivitas Angry Birds the Movie yan...   0        0   
13160  [NOSTALGIA DENGAN BUDAYA]; ; nos·tal·gia n k...   0        0   
13161  brengsek itu orang terbuat dr apa bikin gue be...   1        1   
13162  Kapolda Babi! Biadap dan Bodoh! Gak punya otak...   1        1   
13163  Orang Yahudi, Kristen dan Muslim yang belum pe...   0        0   
13164  USER jangan asal ngomong ndasmu. congor lu yg ...   1        1   
13165                       USER Kasur mana enak kunyuk'   0        1   
13166  USER Hati hati bisu :( .g\n\nlagi bosan huft \...   0        0   
13167  USER USER USER USER Bom yang real mudah terdet...   0        0   
13168  USER Mana situ ngasih(": itu cuma foto ya kuti...   1        1   

       HS_Individual  HS_Group  HS_Religion  HS_Race  HS_Physical  HS_Gender  \
13158              0         0            0        0            0          0   
13159              0         0            0        0            0          0   
13160              0         0            0        0            0          0   
13161              1         0            0        0            0          0   
13162              0         1            0        0            0          0   
13163              0         0            0        0            0          0   
13164              1         0            0        0            1          0   
13165              0         0            0        0            0          0   
13166              0         0            0        0            0          0   
13167              0         0            0        0            0          0   
13168              1         0            0        0            0          0   

       HS_Other  HS_Weak  HS_Moderate  HS_Strong  
13158         0        0            0          0  
13159         0        0            0          0  
13160         0        0            0          0  
13161         1        1            0          0  
13162         1        0            1          0  
13163         0        0            0          0  
13164         0        1            0          0  
13165         0        0            0          0  
13166         0        0            0          0  
13167         0        0            0          0  
13168         1        1            0          0

# ii.a Cek note ekstra cleansing
cek alay_dict dan abuse_dict

In [ ]:
abusive_dict

ABUSIVE
0           alay
1          ampas
2           buta
3        keparat
4         anjing
..           ...
120        rezim
121        sange
122       serbet
123        sipit
124  transgender

[125 rows x 1 columns]

In [ ]:
alay_dict

original                        replacement
0      anakjakartaasikasik           anak jakarta asyik asyik
1             pakcikdahtua                  pak cik sudah tua
2           pakcikmudalagi                  pak cik muda lagi
3              t3tapjokowi                       tetap jokowi
4                       3x                          tiga kali
...                    ...                                ...
15162            mendikbud  menteri pendidikan dan kebudayaan
15163               mendag                menteri perdagangan
15164              menaker               menteri tenaga kerja
15165             memetwit                            mentwit
15166             megangin                           memegang

[15167 rows x 2 columns]

# iv INFORMASI SEMENTARA DARI DATA

**A.INFORMASI YANG DITEMUKAN DARI TABEL DATA CSV**



*   ada 13169 baris × 13 kolom
*   ada kalimat yang dalam nya pakai emot contoh [:p, \x9f \xf0 \x9f\ x98\x82 ]
*   Kata USER ngulang ngulang

**B.INFORMASI YANG DITEMUKAN DARI TABEL CLEANSING**

alay_dict 

*   MEMISAHKAN DAN MENGGANTI KARAKTER YANG RAPAT
*   MENGGUNAKAN HURUF KECIL SEMUA
*   JIKA ADA ANGKA DI GANTI  JADI STRING
*   SINGKATAN LEMBAGA DI UBAH MENJADI PANJANG
*   KARAKTER YANG BERLEBIIHAN DI PERSINGKAT CCONTOH (AAAAMIIIIN = AMIN)

abusive_dict

*   AMAN

# B. BERESIN DAN CLEANSING DATA

kita beresi datanya agar menjadi teks - teks yang lebih masuk akal

harus import Regex

B.1.   DI BERESI FORMATNYA JADI KECIL SEMUA

B.2.   PAKAI DICT UNTUK MERUBAH TEKS

B.3.   PAKAI REGEX HAPUS HAPUSIN YANG KOTOR

B.4.   TEST FUNCTION KEDALAM DATA

In [ ]:
import re

#B.1.   DI BERESI FORMATNYA JADI KECIL SEMUA  
#B.2.   PAKAI DICT UNTUK MERUBAH TEKS



DARI https://www.kaggle.com/code/corneliustantius/indonesia-hatespeech-recognition-main-notebook

* sebelumnya coba buat def satu satu, tetapi tidak bisa di pakai, sekarang coba langkah b.1 b.2 langsung 1 def


https://www.kaggle.com/code/ilhamfp31/simpler-preprocess-indonesian-hate-abusive-text/notebook

In [ ]:
  def lowercase(text):
      return text.lower()

  alay_dict_map = dict(zip(alay_dict['original'], alay_dict['replacement']))
  def normalize_alay(text):
      return ' '.join([alay_dict_map[word] if word in alay_dict_map else word for word in text.split(' ')])

  def remove_nonaplhanumeric(text):
    text = re.sub('[^0-9a-zA-Z]+', ' ', text) 
    return text
  
  def remove_unnecessary_char(text):
    text = re.sub('\n',' ',text) # Remove every '\n'
    text = re.sub('\n\n',' ',text) # Remove every '\n'
    text = re.sub('rt',' ',text) # Remove every retweet symbol
    text = re.sub('user',' ',text) # Remove every username
    text = re.sub('((www\.[^\s]+)|(https?://[^\s]+)|(http?://[^\s]+))',' ',text) # Remove every URL
    text = re.sub('  +', ' ', text) # Remove extra spaces
    return text


uji_v1 = "BEBAAss lah, BABE. bagus ga nih BUAT nG3Tesnya " #BUAT CONTOH KALIMAT

uji_v2 = "aamiin adek abis" #BUAT CONTOH KALIMAT

uji_v3 = "Hehe\n\n rt user user user apa kabs www.google.com\n  hehe" #BUAT CONTOH KALIMAT

uji_v4 = "Halooo,,,,, duniaa!!" #BUAT CONTOH KALIMAT

print("SEBELUM lowercase: ",  (uji_v1), " ==>> " "SESUDAH lowercase: ", lowercase(uji_v1))

print("SEBELUM normalize_alay: ",  (uji_v2), " ==>> " "SESUDAH normalize_alay: ", normalize_alay(uji_v2))

print("SEBELUM remove_unnecessary_char: ",  (uji_v3), " ==>> " "SESUDAH remove_unnecessary_char: ", tweet_cleaner(uji_v3))

print("SEBELUM remove_nonaplhanumeric: ",  (uji_v4), " ==>>" "SESUDAH remove_nonaplhanumeric ", remove_nonaplhanumeric(uji_v4))

SEBELUM lowercase:  BEBAAss lah, BABE. bagus ga nih BUAT nG3Tesnya   ==>> SESUDAH lowercase:  bebaass lah, babe. bagus ga nih buat ng3tesnya 
SEBELUM normalize_alay:  aamiin adek abis  ==>> SESUDAH normalize_alay:  amin adik habis
SEBELUM remove_unnecessary_char:  Hehe

 rt user user user apa kabs www.google.com
  hehe  ==>> SESUDAH remove_unnecessary_char:  Hehe apa kabs hehe
SEBELUM remove_nonaplhanumeric:  Halooo,,,,, duniaa!!  ==>>SESUDAH remove_nonaplhanumeric  Halooo duniaa 


In [ ]:
def preprocess(text):
    text = lowercase(text) # 1
    text = remove_unnecessary_char(text) # 2
    text = normalize_alay(text) # 3
    text = remove_nonaplhanumeric(text)
    return text

uji_v4 = "BAAss lah, BABE. aamiin adek \n rt user user apa kabss www.google.com\n  hehe"

print("SEBELUM preprocess: ",  (uji_v4), " ==>> " "SESUDAH preprocess: ", preprocess(uji_v4))

SEBELUM preprocess:  BAAss lah, BABE. aamiin adek 
 rt user user apa kabss www.google.com
  hehe  ==>> SESUDAH preprocess:  baass lah babe amin adik apa kabss hehe


PAKAI preprocess ITU KEDALAM DATA tweet


In [ ]:
tw_bersih = data['Tweet'].apply(preprocess)
tw_bersih

0         di saat semua cowok berusaha melacak perhatia...
1         siapa yang telat memberi tau elu edan sarap g...
2        41 kadang aku berfikir kenapa aku tetap percay...
3         aku itu aku n nku tau matamu sipit tapi dilih...
4         kaum cebong kafir sudah kelihatan dongoknya d...
                               ...                        
13164     jangan asal berbicara ndasmu congor kamu yang...
13165                              kasur mana enak kunyuk 
13166     hati hati bisu g n nlagi bosan duh xf0 x9f x9...
13167     bom yang real mudah terdeteksi bom yang terku...
13168        mana situ ngasih itu cuma foto ya kutil onta 
Name: Tweet, Length: 13169, dtype: object

In [ ]:
print(tw_bersih)

0         di saat semua cowok berusaha melacak perhatia...
1         siapa yang telat memberi tau elu edan sarap g...
2        41 kadang aku berfikir kenapa aku tetap percay...
3         aku itu aku n nku tau matamu sipit tapi dilih...
4         kaum cebong kafir sudah kelihatan dongoknya d...
                               ...                        
13164     jangan asal berbicara ndasmu congor kamu yang...
13165                              kasur mana enak kunyuk 
13166     hati hati bisu g n nlagi bosan duh xf0 x9f x9...
13167     bom yang real mudah terdeteksi bom yang terku...
13168        mana situ ngasih itu cuma foto ya kutil onta 
Name: Tweet, Length: 13169, dtype: object


In [ ]:
tw_bersih.head(30)

0      di saat semua cowok berusaha melacak perhatia...
1      siapa yang telat memberi tau elu edan sarap g...
2     41 kadang aku berfikir kenapa aku tetap percay...
3      aku itu aku n nku tau matamu sipit tapi dilih...
4      kaum cebong kafir sudah kelihatan dongoknya d...
5      ya bani taplak dan kawan kawan xf0 x9f x98 x8...
6     deklarasi pilihan kepala daerah 2018 aman dan ...
7     gue baru saja selesai re watch aldnoah zero pa...
8     nah admin belanja satu lagi po terbaik nak mak...
9                         enak lagi kalau sambil ngewe 
10    setidaknya gue punya jari tengah buat lu sebel...
11     banci kaleng malu tidak bisa jawab pe anyaan ...
12    kalau belajar ekonomi mestinya jago memprivati...
13    aktor huru hara 98 prabowo si ingin lengserkan...
14     bu guru enakan jadi jablay atau guru sekolah ...
15     lawan bicara gue tidak intelek kayak loe yang...
16          belakangan ini kok pikiran ampas banget ya 
17    ari sama beki adalah rapi monyet xf0 x9f x

#EKSPORT

In [ ]:
tw_bersih.to_csv('preprocessed_tweet_indo_cleansing.csv')

In [ ]:
tw_bersih

0         di saat semua cowok berusaha melacak perhatia...
1         siapa yang telat memberi tau elu edan sarap g...
2        41 kadang aku berfikir kenapa aku tetap percay...
3         aku itu aku n nku tau matamu sipit tapi dilih...
4         kaum cebong kafir sudah kelihatan dongoknya d...
                               ...                        
13164     jangan asal berbicara ndasmu congor kamu yang...
13165                              kasur mana enak kunyuk 
13166     hati hati bisu g n nlagi bosan duh xf0 x9f x9...
13167     bom yang real mudah terdeteksi bom yang terku...
13168        mana situ ngasih itu cuma foto ya kutil onta 
Name: Tweet, Length: 13169, dtype: object